### ISSUE TITLE
Issues with symbolic/infix constructors in data declarations

### ISSUE THREAD

**Issue by sheaf** [JustusAdam/language-haskell#132]
The following causes the extension quite a bit of trouble

```haskell
data C a b = C a b
data a `D` b = a `D` b
data a `E` b where

data (:>) a b = (:>) a b
data a :< b = a :< b
data a := b = a := b
data (:<>) a b where
  (:<>) :: a -> b -> (:<>) a b
data a :>< b where

type F a b = C a b
type a `G` b = a `D` b

type (:>>) a b = a :> b
type a :<< b = a :< b
type a :== b = a := b
```

![symbolic_data_cons_highlighting](https://user-images.githubusercontent.com/1297748/80494427-ef23ed00-8966-11ea-9ef4-3b2a90fbafd7.png)

**Comment by sheaf** [comment]
I think the most important is to get the following two scopes right

- `storage.type.haskell`, applies to non-reserved identifiers in a type context starting with a capital letter (prefix and infix) or any allowed symbol (not necessarily `:`)

```haskell
    data C a b
    data a `D` b
    type E a b = F a b
    type a `G` b = a `H` b
    data a :+: b
    type a :<>: b = a :& b
    type a <.> b = a <!> b
```

- `constant.other.haskell` applies to non-reserved identifiers in a value context starting with a capital letter (prefix or infix) or the symbol `:`

```haskell
data X a b = D a b
data Y a b = (:>) a b
data Z a b = a :& b
```

Less importantly, additional scopes can be added for the infix forms, something along the lines of:

1. `operator.custom.haskell` for symbolic operators in a value context that do not start with `:`,
2. `type.operator.custom.haskell` for any symbolic operator in a type context,
3. `constant.custom.haskell` (or some other suggestion) for symbolic constructors, i.e. non-reserved symbolic identifiers at the value level that start with `:`,
4. `type.operator.infix.haskell` and `constant.infix.haskell` for infix applications of identifiers starting with a capital letter, at the type and value level respectively.

**Comment by sheaf** [comment]
Fixed with 34df8d0 and 7217f36. I'm quite happy as I think this is very helpful for newcomers, to highlight the type and the constructor differently even though they might have the same name.
